File: engine/src/spec.rs
// The subset of spec/ipabet.json's shape the engine actually reads. serde
// handles the parsing generically.

use serde::Deserialize;
use std::collections::HashMap;

#[derive(Deserialize)]
pub struct Spec {
    pub letters: Vec<LetterEntry>,
    pub marks: Vec<MarkEntry>,
    pub superscripts: SupSubTable,
    pub subscripts: SupSubTable,
    #[serde(rename = "optShift", default)]
    pub opt_shift: HashMap<String, String>,
    pub quotes: Quotes,
}

#[derive(Deserialize)]
pub struct LetterEntry {
    pub key: String,
    pub glyph: String,
}

#[derive(Deserialize)]
pub struct MarkEntry {
    pub opt: String,
    pub mark: String,
    #[serde(rename = "type")]
    pub kind: String,
    pub double: Option<String>,
    #[serde(rename = "doubleSpacing", default)]
    pub double_spacing: bool,
    #[serde(default)]
    pub cycle: Vec<String>,
    #[serde(rename = "doubleCycle", default)]
    pub double_cycle: Vec<String>,
    pub clone: Option<String>,
    #[serde(rename = "doubleClone")]
    pub double_clone: Option<String>,
    #[serde(default)]
    pub exclusive: bool,
}

#[derive(Deserialize)]
pub struct SupSubTable {
    pub table: Vec<SupSubEntry>,
}

#[derive(Deserialize)]
pub struct SupSubEntry {
    pub base: String,
    pub sup: Option<String>,
    pub sub: Option<String>,
}

#[derive(Deserialize)]
pub struct Quotes {
    pub default: String,
    pub locales: HashMap<String, Vec<String>>,
}

use roxmltree::{Document, Node};
use std::collections::HashSet;

/// Build a Spec from the LDML source (spec/ipabet.xml). This is the Rust port of
/// spec/tools/ldml-to-spec.ts, proven to reconstruct this Spec so the engine
/// passes its full parity + fuzz suite from the LDML alone.
pub fn parse_ldml(xml: &str) -> Result<Spec, String> {
    let doc = Document::parse(xml).map_err(|e| e.to_string())?;
    let all = |t: &'static str| -> Vec<Node> {
        doc.descendants().filter(|n| n.tag_name().name() == t).collect()
    };

    let mut keys: HashMap<String, String> = HashMap::new();
    for n in all("key") {
        if let Some(id) = n.attribute("id") {
            keys.insert(id.to_string(), n.attribute("output").unwrap_or("").to_string());
        }
    }
    let mut layers: HashMap<String, Vec<Vec<String>>> = HashMap::new();
    for layer in all("layer") {
        let mods = layer.attribute("modifiers").unwrap_or("none").to_string();
        let rows: Vec<Vec<String>> = layer
            .children()
            .filter(|r| r.tag_name().name() == "row")
            .map(|r| r.attribute("keys").unwrap_or("").split_whitespace().map(String::from).collect())
            .collect();
        layers.insert(mods, rows);
    }
    let transforms: Vec<(String, String)> = all("transform")
        .iter()
        .filter_map(|n| Some((n.attribute("from")?.to_string(), n.attribute("to")?.to_string())))
        .collect();
    let mut disp: HashMap<String, String> = HashMap::new();
    for n in all("display") {
        if let (Some(out), Some(d)) = (n.attribute("output"), n.attribute("display")) {
            if let Some(name) = out.strip_prefix("\\m{").and_then(|s| s.strip_suffix('}')) {
                disp.insert(name.to_string(), d.to_string());
            }
        }
    }
    // marker name -> combining char, from each dead-key rule \m{NAME}(.) -> $1\u{XXXX}
    let mut marker_char: HashMap<String, char> = HashMap::new();
    for (f, t) in &transforms {
        if let (Some(name), Some(hx)) = (
            f.strip_prefix("\\m{").and_then(|s| s.strip_suffix("}(.)")),
            t.strip_prefix("$1\\u{").and_then(|s| s.strip_suffix('}')),
        ) {
            if let Some(ch) = u32::from_str_radix(hx, 16).ok().and_then(char::from_u32) {
                marker_char.insert(name.to_string(), ch);
            }
        }
    }

    // letters: 26 single (b_a..b_z) + digraphs (base(\p{M}*)op)
    let mut letters = Vec::new();
    for c in b'a'..=b'z' {
        if let Some(g) = keys.get(&format!("b_{}", c as char)) {
            letters.push(LetterEntry { key: (c as char).to_string(), glyph: g.clone() });
        }
    }
    const PM: &str = "(\\p{M}*)";
    for (f, t) in &transforms {
        if let Some(idx) = f.find(PM) {
            let base = &f[..idx];
            if base.len() == 1 && base.as_bytes()[0].is_ascii_alphanumeric() {
                letters.push(LetterEntry {
                    key: format!("{base}{}", &f[idx + PM.len()..]),
                    glyph: t.strip_suffix("$1").unwrap_or(t).to_string(),
                });
            }
        }
    }

    // marks from the option layers + dead keys + displays + special
    const ROWS: [&str; 4] = ["`1234567890-=", "qwertyuiop[]\\", "asdfghjkl;'", "zxcvbnm,./"];
    let grid = |layer: &str| -> HashMap<char, String> {
        let mut out = HashMap::new();
        if let Some(rws) = layers.get(layer) {
            for (ri, ids) in rws.iter().enumerate() {
                for (ci, id) in ids.iter().enumerate() {
                    if let Some(ch) = ROWS[ri].chars().nth(ci) {
                        out.insert(ch, id.clone());
                    }
                }
            }
        }
        out
    };
    let alt_r = grid("altR");
    let alt_rs = grid("altR shift");
    let char_of = |id: &str| -> Option<char> {
        if let Some(name) = id.strip_prefix("mk_") {
            marker_char.get(name).copied()
        } else if id.starts_with("sp_") {
            keys.get(id).and_then(|s| s.chars().next())
        } else {
            None
        }
    };
    let mut cyc: HashMap<String, Vec<String>> = HashMap::new();
    for n in all("cycle") {
        if let (Some(m), Some(fam)) = (n.attribute("marker"), n.attribute("family")) {
            cyc.insert(m.to_string(), fam.split_whitespace().map(String::from).collect());
        }
    }
    let mut excl: HashSet<String> = HashSet::new();
    for n in all("pair") {
        if let Some(a) = n.attribute("a") {
            excl.insert(a.to_string());
        }
    }
    let fam_chars = |name: &str| -> Vec<String> {
        cyc.get(name)
            .map(|fam| fam.iter().skip(1).filter_map(|n| marker_char.get(n)).map(|c| c.to_string()).collect())
            .unwrap_or_default()
    };

    let mut marks = Vec::new();
    for (phys, id) in &alt_r {
        if !(id.starts_with("mk_") || id.starts_with("sp_")) {
            continue;
        }
        let name = &id[3..];
        let combining = id.starts_with("mk_");
        let mark = match char_of(id) {
            Some(c) => c.to_string(),
            None => continue,
        };
        let mut e = MarkEntry {
            opt: phys.to_string(),
            mark,
            kind: if combining { "combining" } else { "spacing" }.to_string(),
            double: None,
            double_spacing: false,
            cycle: Vec::new(),
            double_cycle: Vec::new(),
            clone: None,
            double_clone: None,
            exclusive: false,
        };
        if let Some(did) = alt_rs.get(phys) {
            if did.starts_with("mk_") || did.starts_with("sp_") {
                e.double = char_of(did).map(|c| c.to_string());
                if did.starts_with("sp_") {
                    e.double_spacing = true;
                }
                if did.starts_with("mk_") {
                    if let Some(d) = disp.get(&did[3..]) {
                        e.double_clone = Some(d.clone());
                    }
                }
                e.double_cycle = fam_chars(&did[3..]);
            }
        }
        if combining {
            if let Some(d) = disp.get(name) {
                e.clone = Some(d.clone());
            }
        }
        e.cycle = fam_chars(name);
        if excl.contains(name) {
            e.exclusive = true;
        }
        marks.push(e);
    }

    // Each \m{raise}<base> -> <sup> (and \m{lower}<base> -> <sub>) transform is
    // one table row; the base carries a regex escape (\( \+) that we strip off.
    let table = |marker: &str, sup: bool| -> SupSubTable {
        let pfx = format!("\\m{{{marker}}}");
        SupSubTable {
            table: transforms
                .iter()
                .filter_map(|(f, t)| {
                    let base = f.strip_prefix(&pfx)?;
                    let base = base.strip_prefix('\\').unwrap_or(base);
                    Some(SupSubEntry {
                        base: base.to_string(),
                        sup: if sup { Some(t.clone()) } else { None },
                        sub: if sup { None } else { Some(t.clone()) },
                    })
                })
                .collect(),
        }
    };
    let superscripts = table("raise", true);
    let subscripts = table("lower", false);

    let mut opt_shift: HashMap<String, String> = HashMap::new();
    for (id, out) in &keys {
        if let Some(d) = id.strip_prefix("os_") {
            if d.len() == 1 && d.as_bytes()[0].is_ascii_digit() {
                opt_shift.insert(d.to_string(), out.clone());
            }
        }
    }

    let mut locales: HashMap<String, Vec<String>> = HashMap::new();
    for n in all("locale") {
        if let Some(id) = n.attribute("id") {
            locales.insert(
                id.to_string(),
                ["open1", "close1", "open2", "close2"].iter().map(|a| n.attribute(*a).unwrap_or("").to_string()).collect(),
            );
        }
    }
    let default = all("quotes").first().and_then(|n| n.attribute("default")).unwrap_or("en").to_string();

    Ok(Spec { letters, marks, superscripts, subscripts, opt_shift, quotes: Quotes { default, locales } })
}

File: js/src/spec.ts
// The spec, parsed from the LDML source (spec/ipabet.xml) at import time — the
// TS counterpart to the engine's Rust `from_ldml`. No build step, no generated
// JSON: the XML is bundled as text and reconstructed here. Both the engine
// (index.ts) and the website read the `spec` object this exports.

// @ts-ignore — bundlers inline this file's text; the XML is the source of truth.
import xml from "../../spec/ipabet.xml";

const unesc = (s: string) =>
  s.replace(/&quot;/g, '"').replace(/&gt;/g, ">").replace(/&lt;/g, "<").replace(/&amp;/g, "&");
const cps = (g: string) => [...g].map((c) => "U+" + c.codePointAt(0)!.toString(16).toUpperCase().padStart(4, "0")).join(" ");
const each = (re: RegExp) => [...xml.matchAll(re)];

// ---- raw XML ----
const keys: Record<string, string> = {};
for (const m of each(/<key id="([^"]+)" output="([^"]*)"\/>/g)) keys[m[1]] = unesc(m[2]);
const layers: Record<string, string[][]> = {};
for (const L of each(/<layer modifiers="([^"]+)">([\s\S]*?)<\/layer>/g))
  layers[L[1]] = [...L[2].matchAll(/<row keys="([^"]+)"\/>/g)].map((r) => r[1].split(/\s+/));
const transforms = each(/<transform from="([^"]+)" to="([^"]*)"\/>/g).map((m) => [unesc(m[1]), unesc(m[2])]);
const disp: Record<string, string> = {};
for (const m of each(/<display output="\\m\{([^}]+)\}" display="([^"]*)"\/>/g)) disp[m[1]] = unesc(m[2]);
// Cycles come from chain transforms \m{P}\m{key} -> \m{R} (R != key): repeated
// presses of `key` walk a family. Follow each key's chain from the key itself.
const cyc: Record<string, string[]> = {};
{
  const edge = new Map<string, string>();
  const keys2 = new Set<string>();
  for (const [f, t] of transforms) {
    const fm = f.match(/^\\m\{([^}]+)\}\\m\{([^}]+)\}$/), tm = t.match(/^\\m\{([^}]+)\}$/);
    if (fm && tm && tm[1] !== fm[2]) { edge.set(fm[1] + "|" + fm[2], tm[1]); keys2.add(fm[2]); }
  }
  for (const key of keys2) {
    const family = [key];
    for (let cur = key; ;) { const r = edge.get(cur + "|" + key); if (r === undefined || r === key) break; family.push(r); cur = r; }
    cyc[key] = family;
  }
}
// Exclusive pairs come from marker-collapse transforms: \m{A}\m{B} -> \m{B}
// means primary A and its double B are one dimension, so B replaces A.
const exclEdges = new Set<string>();
for (const [f, t] of transforms) {
  const fm = f.match(/^\\m\{([^}]+)\}\\m\{([^}]+)\}$/), tm = t.match(/^\\m\{([^}]+)\}$/);
  if (fm && tm && tm[1] === fm[2]) exclEdges.add(fm[1] + "|" + fm[2]);
}

const markerChar: Record<string, string> = {};
for (const [f, t] of transforms) {
  const fm = f.match(/^\\m\{([^}]+)\}\(\.\)$/), tm = t.match(/^\$1\\u\{([0-9A-Fa-f]+)\}$/);
  if (fm && tm) markerChar[fm[1]] = String.fromCodePoint(parseInt(tm[1], 16));
}

// ---- letters ----
interface Letter { key: string; glyph: string; cp: string }
const letters: Letter[] = [];
for (let c = 97; c <= 122; c++) {
  const id = "b_" + String.fromCharCode(c), g = keys[id];
  if (g !== undefined) letters.push({ key: String.fromCharCode(c), glyph: g, cp: cps(g) });
}
for (const [f, t] of transforms) {
  if (!f.includes("(\\p{M}*)")) continue;
  const base = f.slice(0, f.indexOf("("));
  if (!/^[a-z0-9]$/i.test(base)) continue;
  const g = t.replace(/\$1$/, "");
  letters.push({ key: base + f.slice(f.indexOf(")") + 1), glyph: g, cp: cps(g) });
}

// ---- marks ----
const ROWS = ["`1234567890-=", "qwertyuiop[]\\", "asdfghjkl;'", "zxcvbnm,./"];
const grid = (layer: string): Record<string, string> => {
  const out: Record<string, string> = {};
  (layers[layer] ?? []).forEach((ids, ri) => ids.forEach((id, ci) => { if (ci < ROWS[ri].length) out[ROWS[ri][ci]] = id; }));
  return out;
};
const altR = grid("altR"), altRS = grid("altR shift");
const charOf = (id: string) => id.startsWith("mk_") ? markerChar[id.slice(3)] : id.startsWith("sp_") ? keys[id] : undefined;
const fam = (nm: string) => (cyc[nm] ?? []).slice(1).map((n) => markerChar[n]).filter(Boolean);

const marks: any[] = [];
for (const [phys, id] of Object.entries(altR)) {
  if (!/^(mk|sp)_/.test(id)) continue;
  const nm = id.slice(3), combining = id.startsWith("mk_"), ch = charOf(id);
  if (ch === undefined) continue;
  const e: any = { opt: phys, mark: ch, type: combining ? "combining" : "spacing", cp: cps(ch) };
  const did = altRS[phys];
  if (did && /^(mk|sp)_/.test(did)) {
    e.double = charOf(did); e.doubleCp = cps(e.double);
    if (did.startsWith("sp_")) e.doubleSpacing = true;
    if (did.startsWith("mk_") && disp[did.slice(3)]) e.doubleClone = disp[did.slice(3)];
    if (fam(did.slice(3)).length) { e.doubleCycle = fam(did.slice(3)); e.doubleCycleCp = e.doubleCycle.map(cps); }
  }
  if (combining && disp[nm]) e.clone = disp[nm];
  if (fam(nm).length) { e.cycle = fam(nm); e.cycleCp = e.cycle.map(cps); }
  if (did && did.startsWith("mk_") && exclEdges.has(nm + "|" + did.slice(3))) e.exclusive = true;
  marks.push(e);
}

// ---- sup/sub, optShift ----
// Each \m{raise}<base> → <sup> (and \m{lower}<base> → <sub>) transform is one
// table row; the base carries a regex escape (\( \+) that we strip back off.
const pairs = (marker: string, k: string) => transforms
  .filter(([f]) => f.startsWith(`\\m{${marker}}`))
  .map(([f, t]) => ({ base: f.slice(`\\m{${marker}}`.length).replace(/^\\/, ""), [k]: t }));
const superscripts = { table: pairs("raise", "sup") };
const subscripts = { table: pairs("lower", "sub") };
const optShift: Record<string, string> = {};
for (const [id, out] of Object.entries(keys)) { const m = id.match(/^os_(\d)$/); if (m) optShift[m[1]] = out; }

export const spec = { letters, marks, superscripts, subscripts, optShift };
export default spec;
